README.md
# Table of Contents
1. [Introduction](#intro)
    * [Repository Structure](#Repository)
    * [Getting Started](#gettingStarted)
    * [Creative Tag](#creative)
2. [Requirements](#requirements)
3. [Partner Module Overview](#overview)
    * [Configuration](#configuration)
    * [Event Model](#eventModel)
    * [Creating a Partner Module](#creatingPartnerModule)
4. [Utility Libraries](#helpers)
    * [Utils](#utils)
    * [BidRoundingTransformer](#bidRounding)

# <a name='intro'></a>Introduction

<b>Welcome to the Index Exchange Partner Certification Process!</b>

Below you will find everything you need to complete the certification process and be a part of the Header Tag Wrapper!

## <a name='Repository'></a>Repository Structure
* `README.md` - This is the main documentation file which should contain everything you need to complete the certification process. If anything is unclear, please refer to this document first.
* `rhythm-one-htb.js` - This is your partner module file, by default it contains a template divided into multiple sections which need to be completed.
* `rhythm-one-htb-validator.js` - This is the validator file for the configuration object that will be passed into your module.
* `rhythm-one-htb-exports.js` - A file that contains all of the modules exports (i.e. any functions that need to be exposed to the outside world).

## <a name='gettingStarted'></a>Getting Started
1. <b>Complete the rhythm-one-htb.js file</b>
    * rhythm-one-htb.js is where all of your adapter code will live.
    * In order to complete the partner module correctly, please refer to the [Partner Module Overview](#overview) and the [Utility Libraries](#helpers) sections.
    * <b>Please refer to the [Partner Requirements and Guidelines](#requirements) when creating your module. Ensure requirements are met to streamline the review process.</b>
2. <b>Complete the rhythm-one-htb-validator.js file</b>
    * This file is where your partner-specific configurations will need to be validated.
    * Things like type and null checks will be done here.
3. <b>Complete the rhythm-one-htb-exports.js file</b>
    * This file will contain any functions that need to be exported or exposed to the outside world. Things like render functions, custom callbacks, etc. Any legacy render functions will also need to be exposed here. Anything added to the `shellInterface.RhythmOneHtb` will be accessible through `window.headertag.RhythmOneHtb`
4. <b>Submitting for Review</b>
    * Once the module has been verified submit a pull request from the `development-v2`branch to the `master-v2` branch for the Index Exchange team to review. If everything is approved, your adapter will be officially certified!

# <a name='requirements'></a> Partner Requirements & Guidelines
In order for your module to be successfully certified, please refer to the following list of requirements and guidelines. Items under required <b><u>must</b></u> be satisfied in order to pass the certification process. Items under guidelines are recommended for an optimal setup and should be followed if possible.

### General

#### Required
* The only targeting keys that can be set are predetermined by Index. The partner module should not be setting targeting on other keys.
* Must support the following browsers: IE 9+, Edge, Chrome, Safari, and Firefox
* Must run system tests in all supported browsers
* Must not use polyfills

#### Recommended
* Please use our helper libraries when possible. Refer to our [Utility Libraries](#helpers) documentation below. All of the utility functions in this library are designed to be backwards compatible with supported browsers to ease cross-browser compatibility.

### Bid Request Endpoint

#### Required
* Must provide cache busting parameter. Cache busting is required to prevent network caches.
* Partner endpoint domain must be consistent. Load balancing should be performed by the endpoint.
* Your endpoint should support HTTPS request. When wrapper loads in secure pages, all requests need to be HTTPS. If you're unable to provide a secure endpoint, we will not be able to make requests to your ad servers.

#### Recommended
* Your module should support a single request architecture (SRA) which has a capability to send multiple bid requests in a single HTTP request.
* Partner should use AJAX method to make bid requests. Please use our [Network](#network) utility library for making network requests, particularly the `Network.ajax` method. This ensures the requests will work with all browsers.

### Bid Response

#### Required
* Returned bid must be a <b>net</b> bid.
* Pass on bid must be explicit in the response.
* All returned bids shall be in the currency trafficked by the publisher.
* Size must be returned in the response. The sizes must also match the requested size. This size parameter will be validated.
* Encrypted prices must be flagged in the response.

### Pixels & Tracking Beacons

#### Required
* Pixels/beacons must only be dropped only when the partner wins the auction and their ad renders.
* Dropping pixels during auction slows down the execution of auctions and is not allowed by Index Exchange.

### DFP Setup

#### Required
* DFP line items, creatives, and render functions will be set up by the Index Exchange team.
<br><br>

# <a name='overview'></a> Partner Module Overview

## <a name='configuration'></a>Prelude: Configuration & Parcels

The Header Tag Wrapper passes each partner a configuration object that has been configured based on a publisher's website. This object contains all the configuration information required for the wrapper, including partner-specific slot mappings, timeouts, and any other partner-specific configuration.
The partner-specific slot mappings dictate how ad slots on the page will map to partner-specific configuration.
There are 2 concepts to be familiar with when understanding how slots on the page are mapped back to partner-specific configuration. Header Tag Slots, which refers to htSlots and partner-specific configuration, which refers to xSlots in the codebase.
* htSlots - This is an abstraction of the googletag.slot object.
  * These will need to be mapped to xSlots.
* xSlots - These are also an abstraction for partner-specific configuration that will be mapped to htSlots.
  * These represent a single partner-specific configuration.
  * An xSlot is how a partner can map their ad server specific identifiers (placementIDs, siteIDs, zoneIDs, etc) to the `htSlot` object.
  * It can represent a single or multiple sizes.
  * Multiple xSlots can be mapped to the same htSlot.

Example Partner Configuration Mapping
```javascript
{
    "partners": {
        "RhythmOneHtb": {
            "enabled": true,
            "configs": {
                "xSlots": {
                    "xSlot1": {
                        "placementID": "123",
                        "sizes": [ [300, 250], [300, 600] ]
                    },
                    "xSlot2": {
                        "placementID": "345",
                        "sizes": [ [300, 250] ]
                    }
                },
                "mapping": {
                    "htSlotID-1": [ "xSlot1" ],
                    "htSlotID-2": [ "xSlot2" ]
                }
            }
        }
    }
}
```

Based on the mapping defined in the partner config, <i>Parcels</i> will be generated for every xSlot/htSlot pair. Parcels are objects that carry different kinds of information throughout the wrapper. In the context of the adapter, parcels are the input into your adapter. Parcels carry information regarding which slots on the publisher's page need demand. More specifically parcels contain information like xSlot reference, htSlot references, demand (after its been applied by the adapter in parseResponse), etc. Each parcel represents a single combination of an htSlot and an xSlot.

Each parcel is an object in the following form:

```javascript
{
    "partnerId": "RONE",
    "htSlot": {
      "__type__": "HeaderTagSlot"
    },
    "ref": "googletag.Slot",      // reference to the slot on the page, in this example it is a googletag slot
    "xSlotRef": {   // contains the reference to the xSlot object from the partner configuration.
      "placementID": "123",
      "sizes": [ [300, 250], [300, 600] ]
    },
    "xSlotName": "1",
    "requestId": "_fwyvecpA" // generated by the wrapper is used to map the creative back to a piece of demand.
                            // This will show up as targeting under the id key, and is used by the render function.
}
```

These parcels will be fed into both of the functions that your adapter needs to implement:

1. First `generateRequestObj` will need to craft a bid request for the `parcels` that need demand.

2. Second `parseResponse` will take the same parcels and apply the demand (set specific properties in the parcel objects) that is returned from the bid requests to the same parcels and send them back to the wrapper.

## <a name='eventModel'></a> High Level Event Model

1. The Header Tag Wrapper script tag is loaded on the page.
2. Wrapper specific configuration validation is performed.
3. All the partner modules are instantiated.
    * Partner-specific configuration validation is performed - checking that all the required fields are provided and that they are in the correct format.
4. An external request for demand is made to the wrapper. This can be via a googletag display or refresh call, or by other methods depending on the wrapper product in use.
The wrapper requests demand from the partner modules for the required slots (provided in the form of parcels).
    * The wrapper calls `generateRequestObj(returnParcels)` for every partner module.
    * The adapter then crafts and returns a request object based on the parcels (containing slot information) specified.
    * The wrapper then sends out a bid request using the request object.
    * Depending on how the adapter is set up and whether JSONP is supported, a response callback (`adResponseCallback`) is called.
    * The adapter parses the response (`parseResponse`) and attaches the demand to the same returnParcels. It also registers the ad creative with the wrapper's render service.
    * The returnParcels are then sent back to the wrapper.
5. The wrapper applies targeting using the demand from the returnParcels.
6. If the partner wins the auction in the ad server, their creative code will be returned and executed.
    * The creative code contains a call to the wrapper's render function.
7. The partner ad is rendered.

## <a name='creatingPartnerModule'></a> Creating a Partner Module

In this section you will be filling out the rhythm-one-htb.js, rhythm-one-htb-exports.js, and the rhythm-one-htb-validator.js files to create your module.

### Step 0: Config Validation (`rhythm-one-htb-validator.js`)
Before you get started on writing the actual code for your module, you need to figure out what your partner configuration (refer to [Configuration](#configuration)) object will look like. This is crucial because it will determine the input (parcels) to your module's core functions.

Once you have a basic idea of what this will look like, and how you will uniquely identify each slot on your server (via xSlot placementId or other inventory codes) you will need to validate this configuration. This validation will be performed by the wrapper using the `rhythm-one-htb-validator.js` file.

The `rhythm-one-htb-validator.js` file contains a single export, a `partnerValidator` function, that takes in the configuration object that will be fed to your module's constructor (refer to [Configuration](#configuration) for an example layout) and validates it via type checks. The type checks are performed using an external library called `schema-inspector`, for which the documentation can be found here https://github.com/Atinux/schema-inspector.


CHANGES.md
# 2.0.3
- Resolved DOCUMENTATION.md file merge conflict

# 2.0.2
- Updated query string parameter name for GDPR
- Removed optional parameters from rhythm-one-htb-validator.js file

# 2.0.1
- Updated review comments, changed window reference to Browser.topWindow

# 2.0.0
- Initial version, implemented the bidder with the new template

rhythm-one-htb.js
'use strict';

////////////////////////////////////////////////////////////////////////////////
// Dependencies ////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

var Browser = require('browser.js');
var Classify = require('classify.js');
var Constants = require('constants.js');
var Partner = require('partner.js');
var Size = require('size.js');
var SpaceCamp = require('space-camp.js');
var System = require('system.js');
var Network = require('network.js');

var ComplianceService;

var RenderService;

//? if (DEBUG) {
var ConfigValidators = require('config-validators.js');
var PartnerSpecificValidator = require('rhythm-one-htb-validator.js');
var Scribe = require('scribe.js');
var Whoopsie = require('whoopsie.js');
//? }

////////////////////////////////////////////////////////////////////////////////
// Main ////////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

/**
 * Partner module template
 *
 * @class
 */
function RhythmOneHtb(configs) {
    /* =====================================
     * Data
     * ---------------------------------- */

    /* Private
     * ---------------------------------- */

    /**
     * Reference to the partner base class.
     *
     * @private {object}
     */
    var __baseClass;

    /**
     * Profile for this partner.
     *
     * @private {object}
     */
    var __profile;
    var getRMPUrl;

    /* =====================================
     * Functions
     * ---------------------------------- */

    /* Utilities
     * ---------------------------------- */

    /**
     * Generates the request URL and query data to the endpoint for the xSlots
     * in the given returnParcels.
     *
     * @param  {object[]} returnParcels
     *
     * @return {object}
     */
    function __generateRequestObj(returnParcels) {
        /* =============================================================================
         * STEP 2  | Generate Request URL
         * -----------------------------------------------------------------------------
         *
         * Generate the URL to request demand from the partner endpoint using the provided
         * returnParcels. The returnParcels is an array of objects each object containing
         * an .xSlotRef which is a reference to the xSlot object from the partner configuration.
         * Use this to retrieve the placements/xSlots you need to request for.
         *
         * If your partner is MRA, returnParcels will be an array of length one. If your
         * partner is SRA, it will contain any number of entities. In any event, the full
         * contents of the array should be able to fit into a single request and the
         * return value of this function should similarly represent a single request to the
         * endpoint.
         *
         * Return an object containing:
         * queryUrl: the url for the request
         * data: the query object containing a map of the query string paramaters
         *
         * callbackId:
         *
         * arbitrary id to match the request with the response in the callback function. If
         * your endpoint supports passing in an arbitrary ID and returning it as part of the response
         * please use the callbackType: Partner.CallbackTypes.ID and fill out the adResponseCallback.
         * Also please provide this adResponseCallback to your bid request here so that the JSONP
         * response calls it once it has completed.
         *
         * If your endpoint does not support passing in an ID, simply use
         * Partner.CallbackTypes.CALLBACK_NAME and the wrapper will take care of handling request
         * matching by generating unique callbacks for each request using the callbackId.
         *
         * If your endpoint is ajax only, please set the appropriate values in your profile for this,
         * i.e. Partner.CallbackTypes.NONE and Partner.Requesttypes.AJAX. You also do not need to provide
         * a callbackId in this case because there is no callback.
         *
         * The return object should look something like this:
         * {
         *     url: 'http://bidserver.com/api/bids' // base request url for a GET/POST request
         *     data: { // query string object that will be attached to the base url
         *        slots: [
         *             {
         *                 placementId: 54321,
         *                 sizes: [[300, 250]]
         *             },{
         *                 placementId: 12345,
         *                 sizes: [[300, 600]]
         *             },{
         *                 placementId: 654321,
         *                 sizes: [[728, 90]]
         *             }
         *         ],
         *         site: 'http://google.com'
         *     },
         *     callbackId: '_23sd2ij4i1' //unique id used for pairing requests and responses
         * }
         */

        /* ---------------------- PUT CODE HERE ------------------------------------ */
        var queryObj = [];
        var callbackId = System.generateUniqueId();

        /* Change this to your bidder endpoint. */

        var baseUrl;

        /* ------------------------ Get consent information -------------------------
         * If you want to implement GDPR consent in your adapter, use the function
         * ComplianceService.gdpr.getConsent() which will return an object.
         *
         * Here is what the values in that object mean:
         *      - applies: the boolean value indicating if the request is subject to
         *      GDPR regulations
         *      - consentString: the consent string developed by GDPR Consent Working
         *      Group under the auspices of IAB Europe
         *
         * The return object should look something like this:
         * {
         *      applies: true,
         *      consentString: "BOQ7WlgOQ7WlgABABwAAABJOACgACAAQABA"
         * }
         *
         * You can also determine whether or not the publisher has enabled privacy
         * features in their wrapper by querying ComplianceService.isPrivacyEnabled().
         *
         * This function will return a boolean, which indicates whether the wrapper's
         * privacy features are on (true) or off (false). If they are off, the values
         * returned from gdpr.getConsent() are safe defaults and no attempt has been
         * made by the wrapper to contact a Consent Management Platform.
         */

        /* ---------------- Craft bid request using the above returnParcels --------- */

        /* ------- Put GDPR consent code here if you are implementing GDPR ---------- */

        /* -------------------------------------------------------------------------- */
        var gdprConsent = ComplianceService.gdpr.getConsent();
        var privacyEnabled = ComplianceService.isPrivacyEnabled();

        if (gdprConsent && privacyEnabled && typeof gdprConsent === 'object') {
            if (typeof gdprConsent.applies === 'boolean') {
                queryObj.push('gdpr=' + gdprConsent.applies);
            }

            if (gdprConsent.consentString !== '') {
                queryObj.push('gdpr_consent=' + gdprConsent.consentString);
            }
        }

        queryObj.push('domain=' + encodeURIComponent(Browser.topWindow.location.hostname));

        queryObj.push('url=' + encodeURIComponent(Browser.topWindow.location.href));

        baseUrl = getRMPUrl(queryObj, returnParcels);

        return {
            url: Browser.getProtocol() + baseUrl,
            data: '',
            callbackId: callbackId
        };
    }

    function getImp(rp) {
        var c = [];
        for (var i = 0; i < rp.length; i++) {
            if (rp[i].xSlotName) {
                c.push(rp[i].xSlotName);
            }
        }

        return c;
    }

    function getSize(rp) {
        var tw = [];
        var th = [];
        for (var i = 0; i < rp.length; i++) {
            if (rp[i].xSlotRef.sizes) {
                var w = [];
                var h = [];
                for (var j = 0; j < rp[i].xSlotRef.sizes.length; ++j) {
                    w.push(rp[i].xSlotRef.sizes[j][0]);
                    h.push(rp[i].xSlotRef.sizes[j][1]);
                }
                tw.push(w.join('|'));
                th.push(h.join('|'));
            }
        }

        return [tw, th];
    }

    function getMediaTypes(rp) {
        var mt = [];
        for (var i = 0; i < rp.length; i++) {
            mt.push(rp[i].xSlotRef.adType && rp[i].xSlotRef.adType === 'video' ? 'v' : 'd');
        }

        return mt;
    }

    function getFloorPrice(rp) {
        var fp = [];
        for (var i = 0; i < rp.length; i++) {
            fp.push(rp[i].xSlotRef.floor || 0);
        }

        return fp;
    }

    getRMPUrl = function (queryObj, returnParcels) {
        var ref = returnParcels[0].xSlotRef;

        var url = '//tag.1rx.io/rmp/{placementId}/0/{path}?z={zone}';

        try {
            url = url.replace(/\{placementId\}/i, ref.placementId || []);
            url = url.replace(/\{zone\}/i, ref.zone || '1r');
            url = url.replace(/\{path\}/i, ref.path || 'mvo');
        } catch (ex) {
        }

        function p(k, v, d) {
            if (v instanceof Array) {
                v = v.join(d || ',');
            }

            if (typeof v !== 'undefined') {
                queryObj.push(encodeURIComponent(k) + '=' + encodeURIComponent(v));
            }
        }

        function attempt(valueFunction, defaultValue) {
            try {
                return valueFunction();
            } catch (ex) {}

            return defaultValue;
        }

        p('title', Browser.topWindow.document.title);
        p('dsh', Browser.getScreenHeight());
        p('dsw', Browser.getScreenWidth());
        p('tz', System.getTimezoneOffset());

        var dtype = 0;
        if ((/(ios|ipod|ipad|iphone|android)/i).test(Browser.getUserAgent())) {
            dtype = 1;
        } else if ((/(smart[-]?tv|hbbtv|appletv|googletv|hdmi|netcast\.tv|viera|nettv|roku|\bdtv\b|sonydtv|inettvbrowser|\btv\b)/i).test(Browser.getUserAgent())) {
            dtype = 3;
        } else {
            dtype = 2;
        }
        p('dtype', dtype);

        p('flash', attempt(function () {
            var n = Browser.topWindow.navigator;
            var plg = n.plugins;
            var m = n.mimeTypes;
            var t = 'application/x-shockwave-flash';
            var x = Browser.topWindow.ActiveXObject;

            if (plg && plg['Shockwave Flash'] && m && m[t] && m[t].enabledPlugin) {
                return 1;
            }

            if (x) {
                try {
                    if (new Browser.topWindow.ActiveXObject('ShockwaveFlash.ShockwaveFlash')) {
                        return 1;
                    }
                } catch (e) {}
            }

            return 0;
        }, 0));

        var heights = [];
        var widths = [];
        var floors = [];
        var mediaTypes = [];
        var configuredImp = [];

        configuredImp = getImp(returnParcels);
        widths = getSize(returnParcels)[0] || [];
        heights = getSize(returnParcels)[1] || [];

        mediaTypes = getMediaTypes(returnParcels);
        floors = getFloorPrice(returnParcels);

        p('imp', configuredImp || ref.placementId);
        p('w', widths);
        p('h', heights);
        p('floor', floors);
        p('t', mediaTypes);
        p('ht', 'indexExchange');

        url += '&' + queryObj.join('&');

        return url;
    };

    /* -------------------------------------------------------------------------- */

    /* Helpers
     * ---------------------------------- */

    /* =============================================================================
     * STEP 5  | Rendering Pixel
     * -----------------------------------------------------------------------------
     *
    */

    /**
     * This function will render the pixel given.
     * @param  {string} pixelUrl Tracking pixel img url.
     */
    function __renderPixel(pixelUrl) {
        if (pixelUrl) {
            Network.img({
                url: decodeURIComponent(pixelUrl),
                method: 'GET'
            });
        }
    }

    /**
     * Parses and extracts demand from adResponse according to the adapter and then attaches it
     * to the corresponding bid's returnParcel in the correct format using targeting keys.
     *
     * @param {string} sessionId The sessionId, used for stats and other events.
     *
     * @param {any} adResponse This is the bid response as returned from the bid request, that was either
     * passed to a JSONP callback or simply sent back via AJAX.
     *
     * @param {object[]} returnParcels The array of original parcels, SAME array that was passed to
     * generateRequestObj to signal which slots need demand. In this funciton, the demand needs to be
     * attached to each one of the objects for which the demand was originally requested for.
     */

    function __parseResponse(sessionId, adResponse, returnParcels) {
        /* =============================================================================
         * STEP 4  | Parse & store demand response
         * -----------------------------------------------------------------------------
         *
         * Fill the below variables with information about the bid from the partner, using
         * the adResponse variable that contains your module adResponse.
         */

        /* This an array of all the bids in your response that will be iterated over below. Each of
         * these will be mapped back to a returnParcel object using some criteria explained below.
         * The following variables will also be parsed and attached to that returnParcel object as
         * returned demand.
         *
         * Use the adResponse variable to extract your bid information and insert it into the
         * bids array. Each element in the bids array should represent a single bid and should
         * match up to a single element from the returnParcel array.
         *
         */

        /* ---------- Process adResponse and extract the bids into the bids array ------------ */
        var bids = [];

        for (var i = 0; adResponse.seatbid && i < adResponse.seatbid.length; i++) {
            for (var j = 0; adResponse.seatbid[i].bid && j < adResponse.seatbid[i].bid.length; j++) {
                bids.push(adResponse.seatbid[i].bid[j]);
            }
        }

        /* --------------------------------------------------------------------------------- */

        for (var k = 0; k < returnParcels.length; k++) {
            var curReturnParcel = returnParcels[k];
            var headerStatsInfo = {};
            var htSlotId = curReturnParcel.htSlot.getId();
            headerStatsInfo[htSlotId] = {};
            headerStatsInfo[htSlotId][curReturnParcel.requestId] = [curReturnParcel.xSlotName];

            var curBid;

            for (var n = 0; n < bids.length; n++) {
                /**
                 * This section maps internal returnParcels and demand returned from the bid request.
                 * In order to match them correctly, they must be matched via some criteria. This
                 * is usually some sort of placements or inventory codes. Please replace the someCriteria
                 * key to a key that represents the placement in the configuration and in the bid responses.
                 */
                /* ----------- Fill this out to find a matching bid for the current parcel ------------- */
                if (curReturnParcel.xSlotName === bids[n].impid) {
                    curBid = bids[n];
                    bids.splice(n, 1);

                    break;
                }
            }

            /* No matching bid found so its a pass */
            if (!curBid) {
                if (__profile.enabledAnalytics.requestTime) {
                    __baseClass._emitStatsEvent(sessionId, 'hs_slot_pass', headerStatsInfo);
                }
                curReturnParcel.pass = true;

                continue;
            }

            /* ---------- Fill the bid variables with data from the bid response here. ------------ */

            /* Using the above variable, curBid, extract various information about the bid and assign it to
             * these local variables */

            /* The bid price for the given slot */
            var bidPrice = curBid.price;

            /* The size of the given slot */
            var bidSize = [Number(curBid.w), Number(curBid.h)];

            /* The creative/adm for the given slot that will be rendered if is the winner.
             * Please make sure the URL is decoded and ready to be document.written.
             */
            var bidCreative = curBid.adm;

            /* The dealId if applicable for this slot. */
            var bidDealId = curBid.dealid;

            /* Explicitly pass */
            var bidIsPass = bidPrice <= 0;

            /* OPTIONAL: tracking pixel url to be fired AFTER rendering a winning creative.
            * If firing a tracking pixel is not required or the pixel url is part of the adm,
            * leave empty;
            */
            var pixelUrl = '';

            /* --------------------------------------------------------------------------------------- */

            curBid = null;
            if (bidIsPass) {
                //? if (DEBUG) {
                Scribe.info(__profile.partnerId + ' returned pass for { id: ' + adResponse.id + ' }.');
                //? }
                if (__profile.enabledAnalytics.requestTime) {
                    __baseClass._emitStatsEvent(sessionId, 'hs_slot_pass', headerStatsInfo);
                }
                curReturnParcel.pass = true;

                continue;
            }

            if (__profile.enabledAnalytics.requestTime) {
                __baseClass._emitStatsEvent(sessionId, 'hs_slot_bid', headerStatsInfo);
            }

            curReturnParcel.size = bidSize;
            curReturnParcel.targetingType = 'slot';
            curReturnParcel.targeting = {};

            var targetingCpm = '';

            //? if (FEATURES.GPT_LINE_ITEMS) {
            targetingCpm = __baseClass._bidTransformers.targeting.apply(bidPrice);
            var sizeKey = Size.arrayToString(curReturnParcel.size);

            if (bidDealId) {
                curReturnParcel.targeting[__baseClass._configs.targetingKeys.pmid] = [sizeKey + '_' + bidDealId];
                curReturnParcel.targeting[__baseClass._configs.targetingKeys.pm] = [sizeKey + '_' + targetingCpm];
            } else {
                curReturnParcel.targeting[__baseClass._configs.targetingKeys.om] = [sizeKey + '_' + targetingCpm];
            }
            curReturnParcel.targeting[__baseClass._configs.targetingKeys.id] = [curReturnParcel.requestId];
            //? }

            //? if (FEATURES.RETURN_CREATIVE) {
            curReturnParcel.adm = bidCreative;
            if (pixelUrl) {
                curReturnParcel.winNotice = __renderPixel.bind(null, pixelUrl);
            }
            //? }

            //? if (FEATURES.RETURN_PRICE) {
            curReturnParcel.price = Number(__baseClass._bidTransformers.price.apply(bidPrice));
            //? }

            var expiry = 0;
            if (__profile.features.demandExpiry.enabled) {
                expiry = __profile.features.demandExpiry.value + System.now();
            }

            var pubKitAdId = RenderService.registerAd({
                sessionId: sessionId,
                partnerId: __profile.partnerId,
                adm: bidCreative,
                requestId: curReturnParcel.requestId,
                size: curReturnParcel.size,
                price: targetingCpm,
                dealId: bidDealId || null,
                timeOfExpiry: expiry,
                auxFn: __renderPixel,
                auxArgs: [pixelUrl]
            });
            //? if (FEATURES.INTERNAL_RENDER) {
            curReturnParcel.targeting.pubKitAdId = pubKitAdId;
            //? }
        }
    }

    /* =====================================
     * Constructors
     * ---------------------------------- */

    (function __constructor() {
        ComplianceService = SpaceCamp.services.ComplianceService;
        RenderService = SpaceCamp.services.RenderService;

        /* =============================================================================
         * STEP 1  | Partner Configuration
         * -----------------------------------------------------------------------------
         *
         * Please fill out the below partner profile according to the steps in the README doc.
         */

        /* ---------- Please fill out this partner profile according to your module ------------ */
        __profile = {
            partnerId: 'RhythmOneHtb',
            namespace: 'RhythmOneHtb',
            statsId: 'RONE',
            version: '2.0.3',
            targetingType: 'slot',
            enabledAnalytics: {
                requestTime: true
            },
            features: {
                demandExpiry: {
                    enabled: false,
                    value: 0
                },
                rateLimiting: {
                    enabled: false,
                    value: 0
                }
            },

            /* Targeting keys for demand, should follow format ix_{statsId}_id */
            targetingKeys: {
                id: 'ix_rone_id',
                om: 'ix_rone_cpm',
                pm: 'ix_rone_cpm',
                pmid: 'ix_rone_dealid'
            },

            /* The bid price unit (in cents) the endpoint returns, please refer to the readme for details */
            bidUnitInCents: 1,
            lineItemType: Constants.LineItemTypes.ID_AND_SIZE,
            callbackType: Partner.CallbackTypes.NONE,
            architecture: Partner.Architectures.SRA,
            requestType: Partner.RequestTypes.ANY
        };

        /* --------------------------------------------------------------------------------------- */

        //? if (DEBUG) {
        var results = ConfigValidators.partnerBaseConfig(configs) || PartnerSpecificValidator(configs);

        if (results) {
            throw Whoopsie('INVALID_CONFIG', results);
        }
        //? }
        __baseClass = Partner(__profile, configs, null, {
            parseResponse: __parseResponse,
            generateRequestObj: __generateRequestObj
        });
    })();

    /* =====================================
     * Public Interface
     * ---------------------------------- */

    var derivedClass = {
        /* Class Information
         * ---------------------------------- */

        //? if (DEBUG) {
        __type__: 'RhythmOneHtb',
        //? }

        //? if (TEST) {
        __baseClass: __baseClass,
        //? }

        /* Data
         * ---------------------------------- */

        //? if (TEST) {
        profile: __profile,
        //? }

        /* Functions
         * ---------------------------------- */

        //? if (TEST) {
        parseResponse: __parseResponse,
        generateRequestObj: __generateRequestObj
        //? }
    };

    return Classify.derive(__baseClass, derivedClass);
}

////////////////////////////////////////////////////////////////////////////////
// Exports /////////////////////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////

module.exports = RhythmOneHtb;
